Sonar workflow fails for PRs targeting release branches with different module structure

## How `workflow_run` Works

The `.github/workflows/sonar.yaml` workflow uses the `workflow_run` trigger, which has a critical limitation documented by GitHub:

> **This event will only trigger a workflow run if the workflow file exists on the default branch.**

This means:
1. When a PR's "Lint Pull Request" workflow completes, it triggers the Sonar workflow
2. GitHub looks for `sonar.yaml` **on the default branch (`main`)** - not on the PR's branch
3. The Sonar workflow executes **using the workflow file from `main`**
4. However, the workflow checks out the **PR's code** (which may be targeting a different branch)

This creates a mismatch: the workflow from `main` assumes the repository structure of `main`, but it's analyzing code that may have a different structure.

**Key Point:** There is no way to make `workflow_run` use the workflow file from the PR's target branch. It always uses the file from the default branch.

## Problem

The Sonar Cloud analysis fails silently for PRs targeting release branches when the release branch has a different module structure than `main`. This leaves the PR with a perpetually pending "[Kroxylicious Operator] SonarCloud Code Analysis" check.

### Example

PR #4111 targets `release/v0.20` and the Sonar workflow fails with:
```
[ERROR] Could not find the selected project in the reactor: :kroxylicious-kubernetes-parent
```

The `:kroxylicious-kubernetes-parent` module was added in #3818 after the `release/v0.20` branch was created, so it doesn't exist on that branch.

### What Happens

1. PR targets `release/v0.20`
2. Lint workflow completes successfully
3. `workflow_run` triggers Sonar workflow **using the file from `main`**
4. Sonar workflow checks out the PR branch (which has `release/v0.20` structure)
5. Proxy Runtime scan succeeds ✅
6. Operator scan tries to build `:kroxylicious-kubernetes-parent` which doesn't exist ❌
7. Maven build fails, SonarCloud never runs, no status check is posted
8. PR shows perpetually pending check ⏳

## Proposed Solutions

### Option 1: Make the workflow adaptive (Recommended)

Detect which modules exist in the checked-out branch and adjust the build accordingly.

**Implementation:**
```yaml
- name: SonarCloud scan on PR for the Operator
  if: env.SONAR_TOKEN_SET == 'true'
  run: |
    # Detect which modules exist
    PROJECTS=':kroxylicious-operator,:kroxylicious-parent,:kroxylicious-operator-dist'
    if [ -f "kroxylicious-kubernetes/pom.xml" ]; then
      PROJECTS=":kroxylicious-operator,:kroxylicious-kubernetes-parent,:kroxylicious-parent,:kroxylicious-operator-dist"
    fi
    
    mvn --batch-mode \
      --activate-profiles 'ci,!qa' \
      --projects "$PROJECTS" \
      -Derrorprone.skip=true \
      -DskipITs=true \
      -Djapicmp.skip=true \
      -Dsonar.projectName='Kroxylicious Operator' \
      -Dsonar.projectKey='kroxylicious_operator' \
      -Dsonar.scm.revision=${{ github.event.workflow_run.head_sha }} \
      -Dsonar.pullrequest.key=${{ fromJson(steps.get_pr_data.outputs.data).number }} \
      -Dsonar.pullrequest.branch=${{ fromJson(steps.get_pr_data.outputs.data).head.ref }} \
      -Dsonar.pullrequest.base=${{ fromJson(steps.get_pr_data.outputs.data).base.ref }} \
      clean \
      verify \
      org.sonarsource.scanner.maven:sonar-maven-plugin:5.5.0.6356:sonar
  env:
    SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

**Pros:**
- Single workflow file
- Works for all branches automatically
- No code duplication
- Maintains `workflow_run` security boundary
- Automatically adapts to any structural differences

**Cons:**
- Requires bash script logic in workflow
- Need to maintain the detection logic

### Option 2: Conditional steps based on module detection

Use workflow conditionals to run different steps based on what exists.

**Implementation:**
```yaml
- name: Check if kubernetes parent exists
  id: check-k8s-parent
  run: |
    if [ -f "kroxylicious-kubernetes/pom.xml" ]; then
      echo "exists=true" >> $GITHUB_OUTPUT
    else
      echo "exists=false" >> $GITHUB_OUTPUT
    fi

- name: SonarCloud scan (with kubernetes parent)
  if: env.SONAR_TOKEN_SET == 'true' && steps.check-k8s-parent.outputs.exists == 'true'
  run: |
    mvn --batch-mode \
      --projects ':kroxylicious-operator,:kroxylicious-kubernetes-parent,:kroxylicious-parent,:kroxylicious-operator-dist' \
      ...

- name: SonarCloud scan (without kubernetes parent)
  if: env.SONAR_TOKEN_SET == 'true' && steps.check-k8s-parent.outputs.exists == 'false'
  run: |
    mvn --batch-mode \
      --projects ':kroxylicious-operator,:kroxylicious-parent,:kroxylicious-operator-dist' \
      ...
```

**Pros:**
- Clear separation of different scenarios
- Easy to understand which path is taken
- GitHub Actions native conditionals

**Cons:**
- Code duplication (two nearly identical Maven commands)
- More workflow YAML to maintain
- Only handles one specific difference (kubernetes-parent)

### Option 3: Change trigger to `pull_request`

Replace `workflow_run` with direct `pull_request` trigger so the workflow file comes from the PR branch.

**Implementation:**
```yaml
on:
  pull_request:
    types: [opened, synchronize, reopened]
```

**Pros:**
- Uses workflow file from PR branch (matches code structure)
- Simpler trigger mechanism
- No need for PR number artifact passing

**Cons:**
- **Security risk:** Runs in fork context for external PRs, exposing `SONAR_TOKEN` secret
- Would need to use `pull_request_target` with complex conditional secret access
- Current `workflow_run` design provides security isolation
- Not recommended by GitHub for workflows that need secrets

### Option 4: Skip operator scan on failure

Make the operator scan step allow failures gracefully.

**Implementation:**
```yaml
- name: SonarCloud scan on PR for the Operator
  if: env.SONAR_TOKEN_SET == 'true'
  continue-on-error: true
  run: |
    mvn --batch-mode \
      --projects ':kroxylicious-operator,:kroxylicious-kubernetes-parent,:kroxylicious-parent,:kroxylicious-operator-dist' \
      ...
```

**Pros:**
- Simple one-line change
- Won't block PRs

**Cons:**
- Silently skips operator analysis for older branches
- Loses code quality checks for operator on release branches
- Hides real failures too
- Not a real solution

## Recommendation

**Option 1 (Adaptive workflow)** is the best solution because:
- ✅ Handles current and future structural differences automatically
- ✅ No code duplication
- ✅ Maintains security boundary of `workflow_run`
- ✅ Works for all branches without manual maintenance
- ✅ Fails explicitly if Maven can't build (not silently)

The detection logic is simple and can be extended if other structural differences arise in the future.

## Workaround for Current PRs

For PRs currently blocked by this issue (like #4111), a maintainer can:
1. Manually dismiss the required status check, OR
2. Remove it from required checks for that specific PR, OR
3. Wait for this issue to be fixed and re-run the workflow

## Related

- #4111 - Example PR affected by this issue
- #3818 - Commit that added `kroxylicious-kubernetes-parent` module
- #4112 - Related cleanup of unused workflow parameters

--- a/.github/workflows/sonar.yaml
+++ b/.github/workflows/sonar.yaml
@@ -122,9 +122,13 @@ jobs:
       - name: SonarCloud scan on PR for the Operator
         if: env.SONAR_TOKEN_SET == 'true'
         run: |
+          PROJECTS=':kroxylicious-operator,:kroxylicious-parent,:kroxylicious-operator-dist'
+          if [ -f "kroxylicious-kubernetes/pom.xml" ]; then
+            PROJECTS=":kroxylicious-operator,:kroxylicious-kubernetes-parent,:kroxylicious-parent,:kroxylicious-operator-dist"
+          fi
           mvn --batch-mode \
             --activate-profiles 'ci,!qa' \
-            --projects ':kroxylicious-operator,:kroxylicious-kubernetes-parent,:kroxylicious-parent,:kroxylicious-operator-dist' \
+            --projects "$PROJECTS" \
             -Derrorprone.skip=true \
             -DskipITs=true \
             -Djapicmp.skip=true \

